python/ray/data/preprocessors/batch_mapper.py

```diff
@@ -36,19 +39,53 @@ class BatchMapper(Preprocessor):
         >>> preprocessor = BatchMapper(fn)
         >>> preprocessor.transform(ds)  # doctest: +SKIP
         Dataset(num_blocks=1, num_rows=3, schema={X: int64})
+        >>>
+        >>> def fn_numpy(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+        ...     return {"X": batch["X"]}
+        >>> preprocessor = BatchMapper(fn_numpy, batch_format="numpy")
+        >>> preprocessor.transform(ds)  # doctest: +SKIP
+        Dataset(num_blocks=1, num_rows=3, schema={X: int64})
 
     Args:
         fn: The function to apply to data batches.
+        batch_format: The preferred batch format to use in UDF. If not given,
+            we will infer based on the input dataset data format.
     """
 
     _is_fittable = False
 
-    def __init__(self, fn: Callable[["pandas.DataFrame"], "pandas.DataFrame"]):
+    def __init__(
+        self,
+        fn: Callable[
+            [Union["pandas.DataFrame", np.ndarray, Dict[str, np.ndarray]]],
+            Union["pandas.DataFrame", np.ndarray, Dict[str, np.ndarray]],
+        ],
```

Comment:
actually question on the type hint here- is a udf with pandas input to numpy output allowed (or vice versa)?

Reply:
There seems to be some inconsistency between the API for `fn` and the API for `_transform_numpy` and `_transform_pandas`

Reply:
`_transform_numpy` only accepts numpy->numpy
`_transform_pandas` only accepts pandas->pandas

therefore numpy->pandas or pandas->numpy udf should not be allowed with the current apis. Type hint for `fn` should be `Union[Callable[pd.DataFrame, pd.DataFrame], Callable[Union[np.ndarray, Dict[str, np.ndarray]], Union[np.ndarray, Dict[str,np.ndarray]]]]`